exasol_transformers_extension/udfs/models/base_model_udf.py

```diff
@@ -0,0 +1,196 @@
+from abc import abstractmethod
+from typing import Iterator, List, Any, Optional
+import torch
+import pandas as pd
+from exasol_transformers_extension.deployment import constants
+from exasol_transformers_extension.utils import device_management, \
+    bucketfs_operations, dataframe_operations
+
+
+class BaseModelUDF:
+    def __init__(self,
+                 exa,
+                 batch_size,
+                 pipeline,
+                 base_model,
+                 tokenizer,
+                 task_name):
+        self.exa = exa
+        self.batch_size = batch_size
+        self.pipeline = pipeline
+        self.base_model = base_model
+        self.tokenizer = tokenizer
+        self.task_name = task_name
+        self.device = None
+        self.cache_dir = None
+        self.last_loaded_model_key = None
+        self.last_loaded_model = None
+        self.last_loaded_tokenizer = None
+        self.last_created_pipeline = None
+
+    def run(self, ctx):
+        device_id = ctx.get_dataframe(1).iloc[0]['device_id']
+        self.device = device_management.get_torch_device(device_id)
+        ctx.reset()
+
+        while True:
+            batch_df = ctx.get_dataframe(num_rows=self.batch_size, start_col=1)
+            if batch_df is None:
+                break
+
+            predictions_df = self.get_predictions_from_batch(batch_df)
+            ctx.emit(predictions_df)
+
+        self.clear_device_memory()
+
+    def get_predictions_from_batch(self, batch_df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perform separate predictions for each model in the dataframe.
+
+        :param batch_df: A batch of dataframe retrieved from context
+
+        :return: Prediction results of the corresponding batched dataframe
+        """
+        result_df_list = []
+        for model_df in \
+                self.extract_unique_model_dataframes_from_batch(batch_df):
+            for param_based_model_df in \
+                    self.extract_unique_param_based_dataframes(model_df):
+                result_df = self.get_prediction(param_based_model_df)
+                result_df_list.append(result_df)
+
+        result_df = pd.concat(result_df_list)
+        return result_df
+
+    def extract_unique_model_dataframes_from_batch(
+            self, batch_df: pd.DataFrame) -> Iterator[pd.DataFrame]:
+        """
+        Extract unique model dataframes with the same model_name, bucketfs_conn,
+        and sub_dir from the dataframe. If the extracted model is not cached,
```

Comment:
The cache dir likely points to /buckets/... which is read only so a download of a non cached  model likely fails.